rust/opendp/src/traits.rs

```diff
@@ -130,95 +84,274 @@ macro_rules! impl_abs_int {
+/// Casting fails for any value not between Self::MIN_CONSECUTIVE and Self::MAX_CONSECUTIVE.
+pub trait ExactIntCast<TI>: Sized + BoundedInt {
+    fn exact_int_cast(v: TI) -> Fallible<Self>;
+}
+macro_rules! impl_exact_int_cast_from {
+    ($ti:ty, $to:ty) => (impl ExactIntCast<$ti> for $to {
+        #[inline]
+        fn exact_int_cast(v: $ti) -> Fallible<Self> {Ok(From::from(v))}
+    })
+}
+macro_rules! impl_exact_int_cast_try_from {
+    ($ti:ty, $to:ty) => (impl ExactIntCast<$ti> for $to {
+        fn exact_int_cast(v: $ti) -> Fallible<Self> {
+            TryFrom::try_from(v).map_err(|e| err!(FailedCast, "{:?}", e))
+        }
+    })
+}
+// top left
+cartesian!{[u8, u16, u32, u64, u128], impl_exact_int_cast_try_from, impl_exact_int_cast_from, impl_exact_int_cast_from}
+// top right
+cartesian!([u8, u16, u32, u64, u128], [i8, i16, i32, i64, i128], impl_exact_int_cast_try_from, impl_exact_int_cast_try_from, impl_exact_int_cast_from);
+// bottom left
+cartesian!([i8, i16, i32, i64, i128], [u8, u16, u32, u64, u128], impl_exact_int_cast_try_from);
+// bottom right
+cartesian!{[i8, i16, i32, i64, i128], impl_exact_int_cast_try_from, impl_exact_int_cast_from, impl_exact_int_cast_from}
+
+macro_rules! impl_exact_int_cast_int_float {
+    ($int:ty, $float:ty) => (impl ExactIntCast<$int> for $float {
+        fn exact_int_cast(v_int: $int) -> Fallible<Self> {
+            let v_float = v_int as $float;
+            if <$float>::MIN_CONSECUTIVE > v_float || <$float>::MAX_CONSECUTIVE < v_float {
+                fallible!(FailedCast, "exact_int_cast: integer is outside of consecutive integer bounds and may be subject to rounding")
+            } else {
+                Ok(v_float)
             }
         }
+    })
+}
+
+cartesian!([u8, u16, i8, i16], [f32, f64], impl_exact_int_cast_from);
+cartesian!([u64, u128, i64, i128, usize], [f32, f64], impl_exact_int_cast_int_float);
+impl_exact_int_cast_int_float!(u32, f32);
+impl_exact_int_cast_from!(u32, f64);
+impl_exact_int_cast_int_float!(i32, f32);
+impl_exact_int_cast_from!(i32, f64);
+
+cartesian!([usize], [u8, u16, u32, u64, u128, i8, i16, i32, i64, i128], impl_exact_int_cast_try_from);
+
+/// Fallible casting where the casted value rounds towards infinity.
+/// This preserves the invariant that the casted value is gte the original value.
+/// For example, casting a 128_u8 to i8 doesn't saturate to i8::MAX (127), it errors.
+pub trait InfCast<TI>: Sized {
+    fn inf_cast(v: TI) -> Fallible<Self>;
+}
+
+macro_rules! impl_inf_cast_exact {
+    ($ti:ty, $to:ty) => (impl InfCast<$ti> for $to {
+        fn inf_cast(v: $ti) -> Fallible<Self> { ExactIntCast::exact_int_cast(v) }
+    })
+}
+cartesian!([u8, u16, u32, u64, u128, i8, i16, i32, i64, i128], impl_inf_cast_exact);
+
+
+macro_rules! impl_inf_cast_from {
+    ($ti:ty, $to:ty) => (impl InfCast<$ti> for $to {
+        #[inline]
+        fn inf_cast(v: $ti) -> Fallible<Self> { Ok(From::from(v)) }
+    })
+}
+
+pub trait BoundedInt {
```

Comment:
I think this name is still a little cryptic, would suggest ExactIntBounds, ExactIntRange, BoundedExactInt or something like that.